### PR TITLE
Timestamp Replacement

### PR DESCRIPTION
**Pull request by avidan-H**
## Status
- [x] In Progress.

There are also a lot of edits in place to make it work in the build that would need to be removed. There is also some setup that I've included here to get it working that might make more sense to be incorporated as part of the default AMIs.

## Description
Implements timestamp replacement script that is executed by `mitmdump`. After recording a mock file, the file is edited to replace timestamp data with constant values. The next time that a test playbook is executed against that mock, the requests are modified as they are made to replace the same timestamp data that was replaced in the mock file with the same constant values so that the requests being made by the test playbook match the requests in the edited mock file.

## Minimum version of Demisto
- [x] 4.5.0

## Does it break backward compatibility?
   - [x] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by avidan-H** [comment]
@Itay4 I think it would be a good idea to try and run a nightly build (2 builds - the first time would write the edited mocks, and the second would hopefully see improvement in the number of tests that are re-recorded) using the timestamp replacer before merging to see if it's helping in the real scenario.

**Comment by Itay4** [comment]
> @Itay4 I think it would be a good idea to try and run a nightly build (2 builds - the first time would write the edited mocks, and the second would hopefully see improvement in the number of tests that are re-recorded) using the timestamp replacer before merging to see if it's helping in the real scenario.

lets do it

**Comment by glicht** [comment]
@avidan-H what is the status? I see this PR is marked `Draft`.

**Comment by avidan-H** [comment]
@glicht yea I'm still working on it. Pretty sure the timestamp replacement stuff is working as I intended but I'm working on getting a holistic picture to verify that using this will improve results.

File: Tests/mock_server.py [hunk 1]
```diff
@@ -249,6 +252,95 @@ def move_mock_file_to_repo(self, playbook_id, thread_index=0, prints_manager=Non
...
+        prints_manager.add_print_job(
+            f'Test "{playbook_id}" problem_keys: \n{json.dumps(problem_keys, indent=4)}',
+            print,
+            thread_index
+        )
+
+        # is there data in problematic_keys.json that needs whitewashing?
+        prints_manager.add_print_job('checking if there is data to whitewash', print, thread_index)
+        needs_whitewashing = False
+        for _, val in problem_keys.items():
```
Comment (Itay4) [inline Tests/mock_server.py:276]:
```suggestion
        for val in problem_keys.values():
```

File: Tests/mock_server.py [hunk 2]
```diff
@@ -249,6 +252,95 @@ def move_mock_file_to_repo(self, playbook_id, thread_index=0, prints_manager=Non
...
+            if val:
+                needs_whitewashing = True
```
Comment (Itay4) [inline Tests/mock_server.py:278]: break after

File: Tests/mock_server.py [hunk 3]
```diff
@@ -249,6 +252,95 @@ def move_mock_file_to_repo(self, playbook_id, thread_index=0, prints_manager=Non
...
+
+        if problem_keys and needs_whitewashing:
+            mock_file_path = os.path.join(path, get_mock_file_path(playbook_id))
+            cleaned_mock_filepath = mock_file_path.strip('.mock') + '_cleaned.mock'
+            # rewrite mock file with problematic key values replaced
+            command = 'mitmdump -ns ~/timestamp_replacer.py '
+            log_file = os.path.join(path, get_log_file_path(playbook_id, record=True))
+            # Handle proxy log output
+            debug_opt = " >>{} 2>&1".format(log_file) if not self.debug else ''
```
Comment (Itay4) [inline Tests/mock_server.py:287]: use f-string

File: Tests/mock_server.py [hunk 4]
```diff
@@ -249,6 +252,95 @@ def move_mock_file_to_repo(self, playbook_id, thread_index=0, prints_manager=Non
...
+                clean_cmd_output = check_output(self.ami.add_ssh_prefix(split_command, ssh_options='-t'))
+                prints_manager.add_print_job(
+                    f'clean_cmd_output.decode()={clean_cmd_output.decode()}',
+                    print,
+                    thread_index
+                )
+            except CalledProcessError as e:
+                cleaning_err_msg = 'There may have been a problem when filtering timestamp data from the mock file.'
+                prints_manager.add_print_job(cleaning_err_msg, print_error, thread_index)
+                err_msg = 'command `{}` exited with return code [{}]'.format(command, e.returncode)
```
Comment (Itay4) [inline Tests/mock_server.py:306]: f-string boy

File: Tests/mock_server.py [hunk 5]
```diff
@@ -267,15 +359,46 @@ def start(self, playbook_id, path=None, record=False, thread_index=0, prints_man
         # Create mock files directory
         silence_output(self.ami.call, ['mkdir', os.path.join(path, get_folder_path(playbook_id))], stderr='null')
 
-        # Configure proxy server
-        actions = '--server-replay-kill-extra --server-replay' if not record else '--save-stream-file'
-        command = "mitmdump --ssl-insecure --verbose --listen-port {} {}".format(self.PROXY_PORT, actions).split()
-        command.append(os.path.join(path, get_mock_file_path(playbook_id)))
+        # if the keys file doesn't exist, create an empty one
```
Comment (Itay4) [inline Tests/mock_server.py:362]:
```suggestion
```

File: Tests/mock_server.py [hunk 6]
```diff
@@ -267,15 +359,46 @@ def start(self, playbook_id, path=None, record=False, thread_index=0, prints_man
...
 
+        log_file = os.path.join(path, get_log_file_path(playbook_id, record))
         # Handle proxy log output
-        if not self.debug:
-            log_file = os.path.join(path, get_log_file_path(playbook_id, record))
-            command.extend(['>{}'.format(log_file), '2>&1'])
+        debug_opt = " >{} 2>&1".format(log_file) if not self.debug else ''
+
+        # Configure proxy server
+        command = "source .bash_profile && mitmdump --ssl-insecure --verbose --listen-port {} {} {}{}".format(
```
Comment (Itay4) [inline Tests/mock_server.py:397]: add comment why the bash source is needed here

File: Tests/timestamp_replacer.py [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+            print('bad keys loaded\n---------------')
+            print(f'self.query_keys={self.query_keys}')
+            print(f'self.form_keys={self.form_keys}')
+            print(f'self.json_keys={self.json_keys}')
+        else:
+            print('"{}" path doesn\'t exist - no bad keys to set'.format(self.bad_keys_filepath))
+            print('not setting bad keys from file')
+
+
+addons = [TimestampReplacer()]
```
Comment (Itay4) [inline Tests/timestamp_replacer.py:464]: add comment

**Comment by avidan-H** [comment]
I've opened a replacement PR [here](https://github.com/demisto/content/pull/7042) based on the cleaned master so I'm closing this. This is because rebasing this branch on the cleaned-up master was very messy with lots of conflicts needing resolution because I'd merged from master prior to the master branch's being cleaned.

**Pull request closed without merging**